internal/dbsqlc/river_job.sql

```diff
@@ -240,18 +285,23 @@ RETURNING *;
 
 -- name: JobSetStateIfRunning :one
 WITH job_to_update AS (
-    SELECT id
+    SELECT
+      id,
+      @state::river_job_state IN ('retryable'::river_job_state, 'scheduled'::river_job_state) AND metadata ? 'cancel_attempted_at' AS should_cancel
     FROM river_job
     WHERE id = @id::bigint
     FOR UPDATE
 ),
 updated_job AS (
     UPDATE river_job
-    SET errors       = CASE WHEN @error_do_update::boolean        THEN array_append(errors, @error::jsonb) ELSE errors       END,
-        finalized_at = CASE WHEN @finalized_at_do_update::boolean THEN @finalized_at                       ELSE finalized_at END,
+    SET
+        finalized_at = CASE WHEN @finalized_at_do_update::boolean THEN @finalized_at
+          WHEN should_cancel THEN now()
+          ELSE finalized_at END,
+        errors       = CASE WHEN @error_do_update::boolean        THEN array_append(errors, @error::jsonb) ELSE errors       END,
         max_attempts = CASE WHEN @max_attempts_update::boolean    THEN @max_attempts                       ELSE max_attempts END,
         scheduled_at = CASE WHEN @scheduled_at_do_update::boolean THEN @scheduled_at                       ELSE scheduled_at END,
-        state = @state
+        state        = CASE WHEN should_cancel THEN 'cancelled'::river_job_state                           ELSE @state::river_job_state END
```

Comment:
TODO: these changes to `JobSetStateIfRunning` are not tested. Waiting for feedback before adding full coverage for the new bits.

Reply:
j/k I added coverage for this and reworked the query to fix a minor bug I found. It now avoids touching any fields unnecessarily when overriding with a `should_cancel` _except_ for `errors`, which still gets appended to if there's an error to write.